Allow non-default scales on polar axes

dstansby opened the pull request:
## PR Summary
After a long excursion down the rabbit hole of polar Axes transforms, I have emerged to working log-polar axes! Unsurprisingly the fix was trivial - see https://github.com/matplotlib/matplotlib/commit/82d5c5f57e349eed829b3a38bf373781bd6597bc. While I was at it the other commits add some more commenting and tidy up some code. Please leave critical reviews and suggestions, would be good to address them while I still have all this context in my head!

![polar_log](https://user-images.githubusercontent.com/6197628/209722442-e1a5324b-5493-4dbc-961a-41bff1059e66.png)


Based on top of https://github.com/matplotlib/matplotlib/pull/24763

Fixes https://github.com/matplotlib/matplotlib/issues/24383
Fixes https://github.com/matplotlib/matplotlib/issues/12803
Fixes probaby https://github.com/matplotlib/matplotlib/issues/7751
Fixes probably https://github.com/matplotlib/matplotlib/issues/5527


## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

**Documentation and Tests**
- [x] Has pytest style unit tests (and `pytest` passes)
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] New plotting related features are documented with examples.

**Release Notes**
- [ ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [ ] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Create a separate branch for your changes and open the PR from this branch. Please avoid working on `main`.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

oscargus commented on lib/matplotlib/projections/polar.py:
```diff
@@ -20,39 +20,50 @@ class PolarTransform(mtransforms.Transform):
     The base polar transform.
 
     This transform maps polar coordinates ``(theta, r)`` into Cartesian
-    coordinates ``(x, y) = (r * cos(theta), r * sin(theta))`` (but does not
+    coordinates ``(x, y) = (r * cos(theta), r * sin(theta))``
```
It may be nice to get this in math-mode, although not required.

dstansby replied:
👍 done

oscargus commented on lib/matplotlib/projections/polar.py:
```diff
@@ -20,39 +20,50 @@ class PolarTransform(mtransforms.Transform):
-            t *= self._axis.get_theta_direction()
-            t += self._axis.get_theta_offset()
+            theta *= self._axis.get_theta_direction()
+            theta += self._axis.get_theta_offset()
```
Is it possible to add a test that uses these lines?

dstansby replied:
It looks like `_apply_theta_transforms=True` case is not used internally, but as a backwards compatability wrapper for users who used `PolarTransform` before its behaviour was changed. I think adding a meaninful test would be tricky here.

oscargus commented on lib/matplotlib/projections/polar.py:
```diff
@@ -840,9 +864,25 @@ def _set_lim_and_transforms(self):
+        # 4. Scale and translate the cartesian values to Axes coordinates
+        #    (here the origin is moved to the lower left of the Axes)
+        # 5. Move and scale to fill the Axes
+        # 6. Convert from Axes coordinates to Figure coordinates
```
I do not get this at all because of lacking background, but is there a particular reason that the last three terms are within brackets and is that reason already mentioned?

dstansby replied:
At a guess because the bit in brackets is the full affine part of the transform.

oscargus commented:
Also, can you please add a test for #12803? Or is it "obvious" that this is tested by the existing one?

dstansby commented:
> Also, can you please add a test for #12803? Or is it "obvious" that this is tested by the existing one?

I think it's obvious, in that it's the same bug that was causing both #12803 https://github.com/matplotlib/matplotlib/issues/24383 (that I've fixed and added a test to).

> Oh, and possibly the other ones as well. (It seems like "Probably fixes" is interpreted as "fixes", so those will be closed upon merging this with the current statement.)

I've updated the comment so these aren't closed if/when this PR is merged. After a bit more investigation I think this fixes them a bit, but not fully.

oscargus reviewed (approved):
Looks good as far as I can tell!

Unsure about 3.7 or 3.8, so I'll try to get some clarity during the dev-call tonight and then merge accordingly.

efiring commented on lib/matplotlib/tests/test_polar.py:
```diff
@@ -434,3 +434,15 @@ def test_cursor_precision():
+    fig = plt.figure()
+    ax = fig.add_subplot(polar=True)
+
+    ax.set_yscale('log')
+    ax.set_ylim(1, 1000)
```
For clarity, I would use `set_rscale` and `set_rlim`.

tacaswell replied:
```suggestion
    ax.set_rscale('log')
    ax.set_rlim(1, 1000)
```
